website/static/js/nodesPrivacy.js

```diff
@@ -0,0 +1,272 @@
+
+    selectNodes: 'Adjust your privacy settings by checking the boxes below. ' +
+                        '<br><br><b>Checked</b> projects and components will be <b>public</b>.  <br><b>Unchecked</b> components will be <b>private</b>.',
+    addonWarning: {
+        nodesPublic: 'The following projects and components will be made <b>public</b>.',
+        nodesPrivate: 'The following projects and components will be made <b>private</b>.'
+    }
+};
+
+function getNodesOriginal(nodeTree, nodesOriginal) {
+    /**
+     * take treebeard tree structure of nodes and get a dictionary of parent node and all its
+     * children
+     */
+    var i;
+    var nodeId = nodeTree.node.id;
+    nodesOriginal[nodeId] = {
+        public: nodeTree.node.is_public,
+        id: nodeTree.node.id,
+        addons: nodeTree.node.addons,
+        title: nodeTree.node.title,
+        canWrite: nodeTree.node.can_write,
+        changed: false
+    };
+
+    if (nodeTree.children) {
+        for (i in nodeTree.children) {
+            nodesOriginal = getNodesOriginal(nodeTree.children[i], nodesOriginal);
+        }
+    }
+    return nodesOriginal;
+}
+
+function patchNodesPrivacy(nodes) {
+    /**
+     * patches all the nodes in a changed state
+     * uses API v2 bulk requests
+     */
+    var nodesPatch = [];
+    var promise;
+    nodesPatch = nodes.splice(0,10).map(function (node) {
+        return {
+            'type': 'nodes',
+            'id': node.id,
+            'attributes': {
+                'public': node.public
+            }
+        };
+    });
+    return $.ajax({
+        url: API_BASE,
+        type: 'PATCH',
+        dataType: 'json',
+        contentType: 'application/vnd.api+json; ext=bulk',
+        crossOrigin: true,
+        xhrFields: {withCredentials: true},
+        processData: false,
+        data: JSON.stringify({
+            data: nodesPatch
+        })
+    }).done(function (response) {
+        if (nodes.length === 0) {
+            return;
+        }
+        return patchNodesPrivacy(nodes);
+    });
+}
+
+var NodesPrivacyViewModel = function(data, parentIsPublic) {
+    /**
+     * view model which corresponds to nodes_privacy.mako (#nodesPrivacy)
+     *
+     * @type {NodesPrivacyViewModel}
+     */
+    var self = this;
+    var nodesOriginal = {};
+    var treebeardUrl = data.node.api_url  + 'get_node_tree/';
+    self.loadingResults = ko.observable(false);
+    //state of current nodes
+    self.nodesState = ko.observableArray();
+    self.nodesState.subscribe(function(newValue) {
+        m.redraw(true);
+    });
+    //original node state on page load
+    self.nodesOriginal = ko.observableArray();
+    self.nodesOriginal.subscribe(function(newValue) {
+    });
+
+    $('#nodesPrivacy').on('hidden.bs.modal', function () {
+        self.clear();
+    });
+
+    self.loadingResults(true); // enables spinner
+    /**
+     * get node tree for treebeard from API V1
+     */
+    $.ajax({
```

Comment:
Fetching data in the constructor complicates testing and re-use. Instead, do the data in a method (e.g. "fetch"), so that you can do:

```js
var vm = new NodesPrivacyViewModel(data, parentIsPublic);
console.log('Nice, we can test the view model without mocking a request!');
vm.fetch().then(function() {
    console.log('Cool, we can test that the vm is updated at this point!');
});
```